src/v/cluster/topic_table.h

```diff
@@ -24,18 +24,56 @@
 #include <absl/container/node_hash_map.h>
 
 #include <span>
+#include <type_traits>
 
 namespace cluster {
-
-/// Topic table represent all topics configuration and partition assignments.
-/// The topic table is copied on each core to minimize cross core communication
-/// when requesting topic information. The topics table provides an API for
-/// Kafka requests and delta API for controller backend. The delta API allows
-/// backend to wait for changes in topics table. Topics table is update directly
-/// from controller_stm. The table is always updated before any actions related
-/// with topic creation or deletion are executed. Topic table is also
-/// responsible for commiting or removing pending allocations
-///
+/**
+ * ## Overview
+ *
+ * Topic table represent all topics configuration and partition assignments.
+ * The topic table is copied on each core to minimize cross core communication
+ * when requesting topic information. The topics table provides an API for
+ * Kafka requests and delta API for controller backend. The delta API allows
+ * backend to wait for changes in topics table. Topics table is update directly
+ * from controller_stm. The table is always updated before any actions related
+ * with topic creation or deletion are executed.
+ *
+ * Topic table is updated through topic_dispatcher.
+ *
+ *
+ * ## Partition movement state machine
+ *
+ * When partition is being moved from replica set A to replica set B it may
+ * either be finished or cancelled. The cancellation on the other hand may be
+ * reverted. The partition movement state transitions are explained in the
+ * following diagram:
+ *
+ * A - replica set A
+ * B - replica set B
+ *
+ * ┌──────────┐   move           ┌──────────┐       finish         ┌──────────┐
+ * │  static  │                  │  moving  │                      │  static  │
+ * │          ├─────────────────►│          ├─────────────────────►│          │
+ * │    A     │                  │  A -> B  │                      │    B     │
+ * └──────────┘                  └────┬─────┘                      └──────────┘
+ *      ▲                             │                                  ▲
+ *      │                             │                                  │
+ *      │                             │ cancel/force_cancel              │
+ *      │                             │                                  │
+ *      │                             │                                  │
+ *      │                             ▼                                  │
+ *      │                        ┌──────────┐                            │
+ *      │        finish          │  moving  │     revert_cancel          │
```

Comment:
I was considering this, but when it comes to writing the code it would be much more complicated. Handling of finish command already is pretty complex and i didn't want to add to it. Given we introduce the feature in the feature table adding a new command is not really an effort and i think it makes handling a little bit easier.